[WIP] testing first integration tests

knusbaum opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality.

ahmed-mez commented on _integration-tests/validator/validator.go:
```diff
@@ -0,0 +1,472 @@
+
+// assembleTraces2 Reads the traces from the fake agent and returns
+// a map structure in the same format as the "output" field of validation.json files.
+func assembleTraces2(r io.Reader) ([]map[string]interface{}, error) {
```
Curious about the `2` in the function name 😅 What does it mean?

knusbaum replied:
Gah, another leftover that I forgot to clean up.

ahmed-mez commented on _integration-tests/validator/validator.go:
```diff
@@ -0,0 +1,472 @@
+			fmt.Printf("Found span with no parent present: %v\n", s)
+			os.Exit(1)
+		}
+		//parent.Children[s.ID] = s
```
```suggestion
```

ahmed-mez commented on _integration-tests/go.mod:
```diff
@@ -0,0 +1,5 @@
+
+go 1.20
+
+require github.com/davecgh/go-spew v1.1.1
```
Where do we use this?

ahmed-mez commented on integration-tests.sh:
```diff
@@ -0,0 +1,89 @@
+
+## We need to change the default behavior of getting the latest release
+## and instead use the head of the main branch.
+go get github.com/datadog/orchestrion@main
```
Shouldn't we import the local orchestrion to make sure we test the new changes on the branch? We can use `replace` in the go.mod, something like `echo "replace github.com/datadog/orchestrion => <path to the local orchestrion directory> " >> go.mod`

ahmed-mez replied:
See how we do it in parametric-tests in dd-trace-go https://github.com/DataDog/dd-trace-go/blob/8868d37cff43e00caaed80733d55a54c3008a9f0/.github/workflows/parametric-tests.yml#L43

ahmed-mez commented on internal/instrument/instrument.go:
```diff
@@ -303,16 +303,19 @@ func addInFunctionCode(list []dst.Stmt, tc *typechecker.TypeChecker, conf config
 				out = append(out, stmt)
 				appendStmt = false
 				out = append(out, ginMiddleware(stmt))
+				stmt.Decorations().Start.Prepend(dd_instrumented)
```
Not sure about this, we already put these changes between `//dd:startwrap` and `//dd:endwrap`. Why do we need `//dd:instrumented`? And why it was added only to Gin, Echo, Chi and not the others ?

darccio replied:
@ahmed-mez I found out why, check my PR #42. Orchestrion applies two patterns: wrapping and instrumenting. The second one is the right one when we add code when a known statement is found, and the added code isn't actually wrapping the existing statement.

@knusbaum I think our PRs are going to clash. If you want to remove the `dd_instrumented` code, they'll merge fine.

ahmed-mez reviewed (commented):
Looks really good! Let's add a some docs in a readme file under `_integration-tests`.

ahmed-mez commented on _integration-tests/validator/validator.go:
```diff
@@ -0,0 +1,472 @@
+// If valid[vi] is not found, we note that it is missing and start
+// looking for valid[vi+1] from fake[lastfi], where lastfi is the
+// index of the last valid field we found in fake.
+func compare(valid, fake []kv, prefix string, indent int, d *diff) {
```
This is a useful but also a good amount of code to maintain! Have you considered using https://pkg.go.dev/github.com/google/go-cmp/cmp#Diff ?

knusbaum replied:
I think we could manipulate the code to use go-cmp, but we'd still have to do all of the work to restructure the traces, and then prune them. In other words, we might be able to get rid of the compare function, but the rest (and some additional code) would still be required.

darccio commented:
@knusbaum I've a similar set of comments that @ahmed-mez already wrote, so I won't duplicate them. My only addition would be: could it be interesting to document any specifics about these integration tests when adding new contribs?

ahmed-mez commented on .github/workflows/tests.yaml:
```diff
@@ -27,3 +30,41 @@ jobs:
+          #DD_POOL_TRACE_CHECK_FAILURES: true
+          #DD_DISABLE_ERROR_RESPONSES: true
+    steps:
+      - name: Get current date
+        id: date
+        #run: echo "::set-output name=date::$(date +'%Y-%m-%d')"
+        run: echo "date="$(date +'%Y-%m-%d') >> $GITHUB_OUTPUT
+      - uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ steps.date.outputs.date }}
+          restore-keys: |
+            ${{ runner.os }}-go-${{ steps.date.outputs.date }}
+      - name: testdate
+        run: echo ${{ steps.date.outputs.date }}
```
Ok looks we're caching the go build, but don't we just need to set `cache: true` in the config of `actions/setup-go` ?

knusbaum commented on .github/workflows/tests.yaml:
```diff
@@ -27,3 +30,41 @@ jobs:
+        image: ghcr.io/datadog/dd-apm-test-agent/ddapm-test-agent:v1.11.0
+        ports:
+          - 8126:8126
+        env:
+          LOG_LEVEL: DEBUG
+          TRACE_LANGUAGE: golang
+          ENABLED_CHECKS: trace_stall,trace_count_header,trace_peer_service,trace_dd_service
+          #PORT: 8126
+          #DD_SUPPRESS_TRACE_PARSE_ERRORS: true
+          #DD_POOL_TRACE_CHECK_FAILURES: true
+          #DD_DISABLE_ERROR_RESPONSES: true
```
Yes, good idea. This was stuff I meant to clean up.

knusbaum commented on _integration-tests/tests/mux/main.go:
```diff
@@ -0,0 +1,39 @@
+	}
+	r.HandleFunc("/ping", ping).Methods("GET")
+
+	//dd:ignore
+	s := &http.Server{
+		Addr:    ":8080",
+		Handler: r,
+	}
```
I need the server instance so I can call `Shutdown` later (line 36)

knusbaum commented on _integration-tests/tests/net_http/validation.json:
```diff
@@ -0,0 +1,13 @@
+	    "resource": "http.request",
+	    "type": "web",
+	    "metrics": {
+		"_sampling_priority_v1": 1.0
```
No, this was really just here for testing that the validator was working for metrics. I can remove it.

ahmed-mez commented on _integration-tests/validator/validator.go:
@wconti27 @rarguelloF feel free to review this file and suggest the alternative feature that already exists in the test agent.

ahmed-mez reviewed (commented):
LGTM. Let's add a some docs in a readme file under _integration-tests with instruction on how to run them locally.

knusbaum commented:
@ahmed-mez @darccio 
I added documentation about running and writing tests, as well as adding better support for running the tests locally with the script.

I also cleaned up  the github action. I ended up removing the explicit cache since it looks like the setup-go cache is mostly working, and it simplifies the setup.

As far as I'm concerned, this is now 100% ready. Let me know if there is anything outstanding.

darccio reviewed (approved):
![](https://media.giphy.com/media/13zeE9qQNC5IKk/giphy.gif)
